atomic_multiswap/src/test.rs

```diff
@@ -0,0 +1,297 @@
+#![cfg(test)]
+
+use super::*;
+use soroban_sdk::{vec, Address, Env, IntoVal};
+
+mod token {
+    soroban_sdk::contractimport!(file = "../soroban_token_spec.wasm");
+}
+
+use token::{Client as TokenClient, TokenMetadata};
+
+fn create_token_contract(e: &Env, admin: &Address) -> TokenClient {
+    let id = e.register_contract_token();
+    let token = TokenClient::new(e, &id);
+    // decimals, name, symbol don't matter in tests
+    token.init(
+        admin,
+        &TokenMetadata {
+            name: "name".into_val(e),
+            symbol: "symbol".into_val(e),
+            decimals: 7,
+        },
+    );
+    token
+}
+
+fn create_atomic_multiswap_contract(e: &Env) -> AtomicMultiSwapContractClient {
+    AtomicMultiSwapContractClient::new(e, e.register_contract(None, AtomicMultiSwapContract {}))
+}
+
+#[test]
+fn test_atomic_multi_swap() {
+    let env: Env = Default::default();
+    let accounts_a = [
+        AccountSwap {
+            account: Account::random(&env),
+            amount: 2000,
+            min_recv: 290,
+        },
+        AccountSwap {
+            account: Account::random(&env),
+            amount: 3000,
+            min_recv: 350,
+        },
+        AccountSwap {
+            account: Account::random(&env),
+            amount: 4000,
+            min_recv: 301,
+        },
+    ];
+    let accounts_b = [
+        AccountSwap {
+            account: Account::random(&env),
+            amount: 300,
+            min_recv: 2100,
+        },
+        AccountSwap {
+            account: Account::random(&env),
+            amount: 295,
+            min_recv: 1950,
+        },
+        AccountSwap {
+            account: Account::random(&env),
+            amount: 400,
+            min_recv: 2900,
+        },
+    ];
+
+    let token_admin = Account::random(&env);
+
+    let token_a = create_token_contract(&env, &token_admin.address());
+    let token_b = create_token_contract(&env, &token_admin.address());
+    token_a.mint(&token_admin, &accounts_a[0].account.address(), &2000);
+    token_a.mint(&token_admin, &accounts_a[1].account.address(), &3000);
+    token_a.mint(&token_admin, &accounts_a[2].account.address(), &4000);
+
+    token_b.mint(&token_admin, &accounts_b[0].account.address(), &300);
+    token_b.mint(&token_admin, &accounts_b[1].account.address(), &295);
+    token_b.mint(&token_admin, &accounts_b[2].account.address(), &400);
+
+    let contract = create_atomic_multiswap_contract(&env);
+
+    let swap_contract_id = env.register_contract_wasm(atomic_swap::WASM);
+
+    contract.swap(
+        &swap_contract_id,
+        &token_a.contract_id,
+        &token_b.contract_id,
+        &Vec::from_array(&env, accounts_a.clone()),
+        &Vec::from_array(&env, accounts_b.clone()),
+    );
+
+    // check that only 4 swaps are authorized and accounts A[0] and B[1] didn't
+    // authorize anything (and hence the signature can be reused).
```

Comment:
What does this mean that the signature can be reused?

Reply:
It means that the swap didn't happen, so from the account perspective this invocation never happened as well.